

# nf-core/genomeassembler pipeline parameters                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                        
Assemble genomes from long ONT or pacbio HiFi reads                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                        
## Input/output options                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                        
Define where the pipeline should find input data and save output data.                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                        
| Parameter | Description | Type | Default | Required | Hidden |                                                                                                                                                                                                        
|-----------|-----------|-----------|-----------|-----------|-----------|                                                                                                                                                                                               
| `input` | Path to comma-separated file containing information about the samples in the experiment. <details><summary>Help</summary><small>You will need to create a design file with information about the samples in your experiment before running the pipeline. Use
this parameter to specify its location. It has to be a comma-separated file with 3 columns, and a header row. See [usage docs](https://nf-co.re/genomeassembler/usage#samplesheet-input).</small></details>| `string` |  | True |  |                                    
| `outdir` | The output directory where the results will be saved. You have to use absolute paths to storage on Cloud infrastructure. | `string` |  | True |  |                                                                                                         
| `email` | Email address for completion summary. <details><summary>Help</summary><small>Set this parameter to your e-mail address to get a summary e-mail with details of the run sent to you when the workflow exits. If set in your user config file                 
(`~/.nextflow/config`) then you don't need to specify this on the command line for every run.</small></details>| `string` |  |  |  |                                                                                                                                    
| `multiqc_title` | MultiQC report title. Printed as page header, used for filename if not otherwise specified. | `string` |  |  |  |                                                                                                                                   
                                                                                                                                                                                                                                                                        
## Reference genome options                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                        
Reference genome related files and options required for the workflow.                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                        
| Parameter | Description | Type | Default | Required | Hidden |                                                                                                                                                                                                        
|-----------|-----------|-----------|-----------|-----------|-----------|                                                                                                                                                                                               
| `genome` | Name of iGenomes reference. <details><summary>Help</summary><small>If using a reference genome configured in the pipeline using iGenomes, use this parameter to give the ID for the reference. This is then used to build the full paths for all required  
reference genome files e.g. `--genome GRCh38`. <br><br>See the [nf-core website docs](https://nf-co.re/usage/reference_genomes) for more details.</small></details>| `string` |  |  |  |                                                                                
| `fasta` | Path to FASTA genome file. <details><summary>Help</summary><small>This parameter is *mandatory* if `--genome` is not specified. If you don't have a BWA index available this will be generated for you automatically. Combine with `--save_reference` to    
save BWA index for future runs.</small></details>| `string` |  |  |  |                                                                                                                                                                                                  
| `igenomes_ignore` | Do not load the iGenomes reference config. <details><summary>Help</summary><small>Do not load `igenomes.config` when running the pipeline. You may choose this option if you observe clashes between custom parameters and those supplied in      
`igenomes.config`.</small></details>| `boolean` |  |  | True |                                                                                                                                                                                                          
| `igenomes_base` | The base path to the igenomes reference files | `string` | s3://ngi-igenomes/igenomes/ |  | True |                                                                                                                                                  
                                                                                                                                                                                                                                                                        
## Institutional config options                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                        
Parameters used to describe centralised config profiles. These should not be edited.                                                                                                                                                                                    
                                                                                                                                                                                                                                                                        
| Parameter | Description | Type | Default | Required | Hidden |                                                                                                                                                                                                        
|-----------|-----------|-----------|-----------|-----------|-----------|                                                                                                                                                                                               
| `custom_config_version` | Git commit id for Institutional configs. | `string` | master |  | True |                                                                                                                                                                    
| `custom_config_base` | Base directory for Institutional configs. <details><summary>Help</summary><small>If you're running offline, Nextflow will not be able to fetch the institutional config files from the internet. If you don't need them, then this is not a    
problem. If you do need them, you should download the files from the repo and tell Nextflow where to find them with this parameter.</small></details>| `string` | https://raw.githubusercontent.com/nf-core/configs/master |  | True |                                  
| `config_profile_name` | Institutional config name. | `string` |  |  | True |                                                                                                                                                                                          
| `config_profile_description` | Institutional config description. | `string` |  |  | True |                                                                                                                                                                            
| `config_profile_contact` | Institutional config contact information. | `string` |  |  | True |                                                                                                                                                                        
| `config_profile_url` | Institutional config URL link. | `string` |  |  | True |                                                                                                                                                                                       
                                                                                                                                                                                                                                                                        
## Generic options                                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                        
Less common options for the pipeline, typically set in a config file.                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                        
| Parameter | Description | Type | Default | Required | Hidden |                                                                                                                                                                                                        
|-----------|-----------|-----------|-----------|-----------|-----------|                                                                                                                                                                                               
| `version` | Display version and exit. | `boolean` |  |  | True |                                                                                                                                                                                                      
| `publish_dir_mode` | Method used to save pipeline results to output directory. <details><summary>Help</summary><small>The Nextflow `publishDir` option specifies which intermediate files should be saved to the output directory. This option tells the pipeline what
method should be used to move these files. See [Nextflow docs](https://www.nextflow.io/docs/latest/process.html#publishdir) for details.</small></details>| `string` |  |  | True |                                                                                     
| `email_on_fail` | Email address for completion summary, only when pipeline fails. <details><summary>Help</summary><small>An email address to send a summary email to when the pipeline is completed - ONLY sent if the pipeline does not exit                         
successfully.</small></details>| `string` |  |  | True |                                                                                                                                                                                                                
| `plaintext_email` | Send plain-text email instead of HTML. | `boolean` |  |  | True |                                                                                                                                                                                 
| `max_multiqc_email_size` | File size limit when attaching MultiQC reports to summary emails. | `string` | 25.MB |  | True |                                                                                                                                           
| `monochrome_logs` | Do not use coloured log outputs. | `boolean` |  |  | True |                                                                                                                                                                                       
| `hook_url` | Incoming hook URL for messaging service <details><summary>Help</summary><small>Incoming hook URL for messaging service. Currently, MS Teams and Slack are supported.</small></details>| `string` |  |  | True |                                          
| `multiqc_config` | Custom config file to supply to MultiQC. | `string` |  |  | True |                                                                                                                                                                                 
| `multiqc_logo` | Custom logo file to supply to MultiQC. File name must also be set in the MultiQC config file | `string` |  |  | True |                                                                                                                               
| `multiqc_methods_description` | Custom MultiQC yaml file containing HTML including a methods description. | `string` |  |  |  |                                                                                                                                       
| `validate_params` | Boolean whether to validate parameters against the schema at runtime | `boolean` | True |  | True |                                                                                                                                               
| `pipelines_testdata_base_path` | Base URL or local path to location of pipeline test dataset files | `string` | https://raw.githubusercontent.com/nf-core/test-datasets/ |  | True |                                                                                  
                                                                                                                                                                                                                                                                        
## Other parameters                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                        
| Parameter | Description | Type | Default | Required | Hidden |                                                                                                                                                                                                        
|-----------|-----------|-----------|-----------|-----------|-----------|                                                                                                                                                                                               
| `ont` | ONT reads available? | `boolean` |  |  |  |                                                                                                                                                                                                                   
| `hifi` | HiFi reads available? | `boolean` |  |  |  |                                                                                                                                                                                                                 
| `short_reads` | Short reads available? | `boolean` |  |  |  |                                                                                                                                                                                                         
| `collect` | collect ONT reads into a single file | `boolean` |  |  |  |                                                                                                                                                                                               
| `porechop` | run porechop on ONT reads | `boolean` |  |  |  |                                                                                                                                                                                                         
| `lima` | run lima on HiFi reads? | `boolean` |  |  |  |                                                                                                                                                                                                               
| `pacbio_primers` | file containing pacbio primers for trimming with lima | `string` |  |  |  |                                                                                                                                                                        
| `trim_short_reads` | trim short reads with trimgalore | `boolean` |  |  |  |                                                                                                                                                                                          
| `assembler` | Assembler to use. Valid choices are: `'hifiasm'`, `'flye'`, or `'flye_on_hifiasm'`. `flye_on_hifiasm` will scaffold flye assembly (ont) on hifiasm (hifi) assembly using ragtag | `string` |  |  |  |                                                   
| `kmer_length` | kmer length to be used for jellyfish | `integer` |  |  |  |                                                                                                                                                                                           
| `read_length` | read length for genomescope (ONT only) | `string` |  |  |  |                                                                                                                                                                                          
| `dump` | dump jellyfish output | `boolean` |  |  |  |                                                                                                                                                                                                                 
| `meryl_k` | kmer length for meryl | `integer` |  |  |  |                                                                                                                                                                                                              
| `use_ref` | use reference genome | `boolean` |  |  |  |                                                                                                                                                                                                               
| `genome_size` | expected genome size | `string` |  |  |  |                                                                                                                                                                                                            
| `flye_mode` | flye mode | `string` | "--nano-hq" |  |  |                                                                                                                                                                                                              
| `flye_args` | additional args for flye | `string` | "" |  |  |                                                                                                                                                                                                        
| `qc_reads` | Long reads that should be used for QC when both ONT and HiFi reads are provided. Options are `'ONT'` or `'HIFI'` | `string` | "ONT" |  |  |                                                                                                              
| `hifiasm_ont` | Use hifi and ONT reads with `hifiasm --ul` | `boolean` |  |  |  |                                                                                                                                                                                     
| `hifiasm_args` | Extra arguments passed to `hifiasm` | `string` | "" |  |  |                                                                                                                                                                                          
| `polish_pilon` | Polish assembly with pilon? | `boolean` |  |  |  |                                                                                                                                                                                                   
| `polish_medaka` | Polish assembly with medaka (ONT only) | `boolean` |  |  |  |                                                                                                                                                                                       
| `medaka_model` | model to use with medaka | `string` | 'r1041_e82_400bps_hac_v4.2.0' |  |  |                                                                                                                                                                          
| `scaffold_ragtag` | Scaffold with ragtag (requires reference)? | `boolean` |  |  |  |                                                                                                                                                                                 
| `scaffold_links` | Scaffolding with links? | `boolean` |  |  |  |                                                                                                                                                                                                     
| `scaffold_longstitch` | Scaffold with longstitch? | `boolean` |  |  |  |                                                                                                                                                                                              
| `lift_annotations` | Lift-over annotations (requires reference)? | `boolean` |  |  |  |                                                                                                                                                                               
| `busco` | Run BUSCO? | `boolean` |  |  |  |                                                                                                                                                                                                                           
| `busoc_db` | Path to busco db | `string` | '' |  |  |                                                                                                                                                                                                                 
| `busco_lineage` | Busco lineage to use | `string` | "brassicales_odb10" |  |  |                                                                                                                                                                                       
| `quast` | Run quast | `boolean` |  |  |  |                                                                                                                                                                                                                            
| `skip_assembly` | skip assembly steps <details><summary>Help</summary><small>Skip assembly and perform only qc.</small></details>| `boolean` |  |  |  |                                                                                                               
| `skip_alignments` | skip alignments during qc | `boolean` |  |  |  |                                                                                                                                                                                                  
| `jellyfish` | run jellyfish on ONT reads to compute k-mer distribution and estimate genome size | `boolean` |  |  |  |                                                                                                                                                
| `yak` | run qc via yak | `boolean` |  |  |  |                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                        


